### PR TITLE
[WIP] Improve management of + and " on OSX

### LINKED ISSUE #3067: Copy/pasting paragraphs in visual mode inserts extra line.
Using v}y}p to copy a paragraph and insert it after itself in Neovim works differently than in Vim. In Neovim an extra line is insert.

Before: http://imgur.com/iMmCxai

Highlight, copy, and paste the 5 lines starting with line 1168: v}y}p

After Vim: http://imgur.com/wU8bHfj

After nvim: http://imgur.com/kS54g5P

### PR DESCRIPTION
**Pull request by mhinz**
Previously `get_clipboard()` was guessing whether to paste char-wise or
line-wise. Now this information is provided by the clipboard provider.

If the clipboard provider doesn't have any informtion about the used regtype
(`v`, `V`, `<c-v>`), e.g. just after starting Vim, it defaults to char-wise.

---

Is supposed to fix: #3067

### CODE REVIEW

**Comment by mhinz** [comment]
When testing manually everything seems to be working (on OSX, therefore `pbcopy`/`pbpaste`), but the tests fail with SIGPIPE. I'm looking into it later.

File: runtime/autoload/provider/clipboard.vim [hunk 1]
```diff
@@ -88,5 +81,11 @@ function! s:clipboard.set(lines, regtype, reg)
 endfunction
 
 function! provider#clipboard#Call(method, args)
-  return call(s:clipboard[a:method],a:args,s:clipboard)
+  if a:method == 'get'
+    let data = s:selections[a:args[0]].data  " empty or [lines, regtype]
+    let regtype = get(data, 1, 'v')
+    return [call(s:clipboard[a:method], a:args, s:clipboard), regtype]
```
Comment (bfredl) [inline runtime/autoload/provider/clipboard.vim]: why split `s:clipboard.get` in half like this? It seem completely arbitarary to me.

File: runtime/autoload/provider/clipboard.vim [hunk 2]
```diff
@@ -30,13 +30,11 @@ function! s:try_cmd(cmd, ...)
   return out
 endfunction
 
-let s:cache_enabled = 1
 if executable('pbcopy')
   let s:copy['+'] = 'pbcopy'
   let s:paste['+'] = 'pbpaste'
   let s:copy['*'] = s:copy['+']
   let s:paste['*'] = s:paste['+']
-  let s:cache_enabled = 0
```
Comment (bfredl) [inline runtime/autoload/provider/clipboard.vim]: it is true that this safeguard isn't really needed, but it serves as documentation that pbcopy doesn't support caching. We should at least have a note here or something.

**Comment by bfredl** [comment]
I agree with moving logic from c to the provider, if it makes things simpler, but there is a slight regression: First copy and paste something charwise locally in nvim1 with `*` register. Then select something linewise in nvim2 (or in gedit or terminal or whatever), and paste in nvim1, this (incorrectly) pastes charwise.

**Comment by bfredl** [comment]
I think I see the tradeoff here. On __OS X__ this fixes full modewise selection _locally_ but OTOH it it breaks the (limited) linewise support _between_ instances. A complete fix would be to implement a proper clipboard provider for OSX, as I suggested in #2077 someone developing on OSX could port https://github.com/vim/vim/blob/master/src/os_macosx.m to nvim (make it a utility or just run it inprocess in nvim)

File: runtime/autoload/provider/clipboard.vim (hunk 1 above)
Comment (mhinz) [inline runtime/autoload/provider/clipboard.vim]: You mean the additional list around the return value? I wanted to have the `[lines, regtype]` format, so the handling in the C code would be easier.

File: runtime/autoload/provider/clipboard.vim (hunk 2 above)
Comment (mhinz) [inline runtime/autoload/provider/clipboard.vim]: I didn't even know these tools support caching in any way..

**Comment by mhinz** [comment]
I thought this to be easier than it is. I didn't know about #2077 either. I'll read up on this before going any further.

Anyway, this is a essential feature broken and therefore should belong to the 0.1 milestone. (cc @justinmk)

File: runtime/autoload/provider/clipboard.vim (hunk 1 above)
Comment (bfredl) [inline runtime/autoload/provider/clipboard.vim]: No I meant why move this at all from `s:clipboard.get`? This function `provider#clipboard#Call` should just be  wrapper to convert autoload call to dict method call.

File: runtime/autoload/provider/clipboard.vim (hunk 2 above)
Comment (bfredl) [inline runtime/autoload/provider/clipboard.vim]: I was a bit unclear... The X tools do no caching, but they enable us to cache clipboard contents, as their process lifetime track ownership, see #2703. OS X clipboard work different so the same model can't be used.

**Comment by bfredl** [comment]
Note that parts of #2077 was broken off and merged, I should update the description...

**Comment by mhinz** [comment]
Okay, apparently I never knew how X is really working. Let me put it in a nutshell..

If I copy something, _Program 1_ takes ownership of the CLIPBOARD selection. Now when I switch to _Program 2_ and try pasting, it requests the data (in a certain format) from the current owner of the clipboard. It's basically a simple form of IPC as in X providing the protocol for 2 processes to communicate with each other.

In OSX, the clipboard is an acual place in memory that serves as shared memory for all interested programs. `pbcopy` and `pbpaste` and really just simple wrappers around the OS's syscalls for managing the "pasteboard" (https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/PasteboardGuide106/Articles/pbCopying.html#//apple_ref/doc/uid/TP40008102-SW1).

And now, instead of going the way of external processes + VimL provider, you suggest implementing the management of that pasteboard directly by porting `os_macosx.m`.

Is this correct?

**Comment by bfredl** [comment]
Yes, either turn `os_macosx.m` into a helper utility like `pbcopy` or just include it in the `nvim` build. I don't know the build situation on osx (will a `.m` source complicate everything, etc) so I don't know what's simplest.

**Comment by fwalch** [comment]
Hm.. to avoid a dependency on X11/Cocoa (which is the reason we ripped out all this clipboard code in the first place, right?), we would need these standalone tools (modified xclip, OSX helper) to be installable separately. For example, you would have e.g. a `neovim-x11-clipboard` Arch Linux package, much like the `python-neovim` package for the Python plugin (or, well, instead of the current `xclip` dependency). Soo.. probably these utilities should be completely separate from the `nvim` build, maybe even with a different repo? To avoid having them in PATH, we could install them in `$VIMRUNTIME/tools` or something.

**Comment by fwalch** [comment]
Some thoughts, correct me if I'm wrong: On Linux, there might be the situation that X11 is not installed, that's why we need a separate clipboard helper (modified `xclip`). On Windows and OSX, there is always a GUI, so we can just compile/link the clipboard helper code with the main `nvim` executable? If https://github.com/kfish/xsel/pull/8 is not merged, we can package this modified version (e.g. `neovim-x11-clipboard` for Arch Linux, replacing the current `xclip` dependency) and install it in `$VIMRUNTIME/tools` or something to not clutter PATH (even if it is merged, we probably need to package xclip e.g. in the Ubuntu PPA so that the updated xclip is available to Neovim users).

**Comment by justinmk** [comment]
@fwalch That's an accurate summary. 

Note that the 'nix situation is not a reason to give up and link to X. It's far more important that users never ever need to decide which "version" of neovim to install. There's always only one version available, and it works with and without X/Wayland/etc.

**Comment by mhinz** [comment]
So incorperating the `.m` was easy, but I wonder if there's a way to avoid these 2 name clashes:

```
/usr/include/MacTypes.h:309:41: error: typedef redefinition with different types ('unsigned char' vs 'bool')
typedef unsigned char                   Boolean;
                                        ^
../src/nvim/api/private/defs.h:40:14: note: previous definition is here
typedef bool Boolean;
             ^
```
and
```
/System/Library/Frameworks/CoreServices.framework/Frameworks/CarbonCore.framework/Headers/Finder.h:244:41: error: typedef redefinition with different types ('struct FileInfo' vs 'struct FileInfo' (aka 'FileInfo'))
typedef struct FileInfo                 FileInfo;
                                        ^
../src/nvim/os/fs_defs.h:9:3: note: previous definition is here
} FileInfo;
  ^
```

**Comment by justinmk** [comment]
@mhinz is this still relevant?

**Comment by mhinz** [comment]
@justinmk It is. I'm looking into it (again).

**Comment by mhinz** [comment]
I was wrong. The behaviour is entirely fine on master!

**Pull request closed without merging**